crates/merkle-inclusion/src/lib.rs

```diff
@@ -0,0 +1,158 @@
+pub mod traits;
+
+use crate::traits::GadgetDigest;
+use bellpepper_core::boolean::Boolean;
+use bellpepper_core::{ConstraintSystem, SynthesisError};
+use ff::PrimeField;
+
+// HashValue represents the digest type output from the hash function
+pub type HashValue = Vec<Boolean>;
+pub type Key = Vec<Boolean>;
+const HASH_LENGTH: usize = 32 * 8;
+
+/// Represents a leaf node in a Jellyfish Merkle Tree.
+///
+/// Each leaf node contains a key-value pair, where the key is used to determine the position in the tree and the value
+/// is the data stored.
+pub struct Leaf {
+    key: Key,
+    value_hash: HashValue,
+}
+
+impl Leaf {
+    pub fn new(key: Key, value_hash: HashValue) -> Self {
+        Self { key, value_hash }
+    }
+    pub fn hash(&self) -> &HashValue {
+        &self.value_hash
+    }
+
+    pub fn key(&self) -> &HashValue {
+        &self.key
+    }
+}
+
+/// Contains a leaf node and sibling hashes required for proof verification in a Merkle Tree.
+///
+/// The `Proof` struct is essential for the Merkle Tree's cryptographic verification process, providing the necessary
+/// information to verify the integrity and authenticity of the data.
+pub struct Proof {
+    leaf: Leaf,
+    siblings: Vec<HashValue>,
+}
+
+impl Proof {
+    pub fn new(leaf: Leaf, siblings: Vec<HashValue>) -> Self {
+        Self { leaf, siblings }
+    }
+
+    pub fn leaf(&self) -> &Leaf {
+        &self.leaf
+    }
+
+    pub fn siblings(&self) -> &[Vec<Boolean>] {
+        &self.siblings
+    }
+}
+
+/// Verifies the Merkle Tree proof against an expected root hash.
+///
+/// This function reconstructs the root hash from the given leaf and sibling hashes and compares it against the provided
+/// expected root hash.
+///
+/// # Arguments
+/// * `cs` - A mutable reference to the constraint system.
+/// * `expected_root` - The expected root hash of the Merkle Tree.
+/// * `proof` - The proof containing the leaf and its sibling hashes.
+///
+/// # Returns
+/// A result containing the reconstructed root hash if successful, or a `SynthesisError` otherwise.
+pub fn verify_proof<E, CS, GD>(
+    mut cs: CS,
+    expected_root: Vec<Boolean>,
+    proof: Proof,
+) -> Result<Vec<Boolean>, SynthesisError>
+where
+    E: PrimeField,
+    CS: ConstraintSystem<E>,
+    GD: GadgetDigest<E>,
+{
+    assert_eq!(expected_root.len(), GD::output_size());
+
+    //Assert that we do not have more siblings than the length of our hash (otherwise cannot know which path to go)
```

Comment:
This is materializing an undocumented assumption that we use an index of size `HASH_LENGTH` to thread the sibling hashes. 

We use the leaf's key to thread sibling hashes.

1. Is there an undocumented assumption on the length of the key here?
2. what if `proof.leaf().key().len() > HASH_LENGTH`? what breaks in the algorithm.

Reply:
So as you have seen, [we expect the key to be smaller or of same size as the output size of our hashing algorithm](https://github.com/lurk-lab/bellpepper-gadgets/pull/16/files#diff-260ab9bebaec6c9babe2300c428142d465ac60b357284985399dd94799bac24aR103). But you are right that we are missing a check here. I will add an assert and also add more tests to ensure that those edge cases are covered.

Reply:
> we expect the key to be smaller or of same size as the output size of our hashing algorithm

Right, that's correct, the code expects the key to be at most as long as our hash output is.

The question I'm asking is: do we need to enshrine *that* assumption in the algorithm, and if so, why?

One *related* assumption that I am sure we need to make is that the key needs to be as long as the siblings size. The reason we need to make that assumption is that otherwise we can't index into the siblings array using the key.

The two assumptions are related, and the later assumption is clearly needed. Is the earlier assumption needed as well?

Reply:
I'm sorry for the answer, I think I wrote it down without thinking it through at first.

To properly navigate our proof and ensure that it corresponds to a valid leaf we need to only need to ensure that `Len(proof.siblings) <= Len(proof.leaf.key)`. It verifies that we will get information on position for all of our siblings.

So you are right to say that the first assumption is not needed and for the second one I would slightly change it to accepts longer keys as it would not break out algorithm anyway.

Reply:
Fixing the iteration process done in 775a5fa.

Assertion for number of siblings and length of the key done in 1280301.